### PR TITLE
test(config): added controller

### LINKED ISSUE #94: Configuration Controller – Validate configuration updates and retrievals (Medium)

### PR DESCRIPTION
**Pull request by tirirote**
Closes  #94

### CODE REVIEW

File: src/controllers/configuration.controller.js [hunk 1]
```diff
@@ -43,13 +43,15 @@ export async function updateConfiguration(req, res) {
...
       },
     });
-    const configId = configuration.dataValues.id;
 
-    if (configuration.length === 0) {
+    if (!configuration) {
       return res
         .status(404)
-        .json({ message: `Configuration ${configId} not found` });
+        .json({ message: `Configuration with endpoint ${endpoint} not found` }); // Mejoramos el mensaje
```
Comment (alvarobernal2412) [inline src/controllers/configuration.controller.js:50]: Remove this comment (should be in English)

File: src/controllers/configuration.controller.js [hunk 2]
```diff
@@ -61,7 +63,7 @@ export async function updateConfiguration(req, res) {
         },
       }
     );
-
+    console.log('About to call updateConfigurationsCache');
```
Comment (alvarobernal2412) [inline src/controllers/configuration.controller.js:66]: The code should not include console.logs

File: src/controllers/configuration.controller.js
Comment (alvarobernal2412) [inline src/controllers/configuration.controller.js:1]: Thank you for the refactorization! The thing is that there are a few aspects that could be improved, you can try asking an AI tool to suggest code improvements.

File: src/utils/dates.js
Comment (alvarobernal2412) [inline src/utils/dates.js:1]: This file does not correspond to this PR

**Review by alvarobernal2412 — CHANGES_REQUESTED** [review]
Here are some areas for improvement in the testing of the controller:  
### **1. Improve Mocking of `updateConfigurationsCache`**  
- The test for `updateConfiguration` does not correctly invoke `updateConfigurationsCache`.  
- The function is referenced but never executed (`await updateConfigurationsCache;` should be `await updateConfigurationsCache();`).  
- The test should ensure this function is actually called.

### **2. Validate Request Data Types**  
- In `updateAssistantLimit`, `req.params.limit` is a string, but it should be validated as a number.  
- Convert and validate `limit` before performing operations:  
  ```js
  const limit = Number(req.params.limit);
  if (isNaN(limit)) { /* return 400 error */ }
  ```
- The tests should check for invalid input cases (e.g., `"abc"` as a limit).

### **3. Add Tests for Edge Cases**  
- **Empty Database Cases**:  
  - What if `findAll()` in `getConfiguration` returns an empty array?  
  - The test cases assume configurations always exist.  
- **Boundary Values for `limit`**:  
  - Test cases should check limits of `1`, `2`, `Number.MAX_SAFE_INTEGER`, etc.  
- **Case Sensitivity**:  
  - What if an endpoint is requested in a different case (`/API/TEST` vs `/api/test`)?  

### **4. Test for Unexpected Input Scenarios**  
- **Invalid JSON Body**:  
  - What happens if the request body is malformed or missing required fields?  
  - Example test:  
    ```js
    mockReq = { body: {} }; 
    await configurationController.updateConfiguration(mockReq, mockRes);
    expect(mockRes.status).toHaveBeenCalledWith(400);
    ```  
- **Null or Undefined Parameters**:  
  - Ensure tests cover cases where parameters (`endpoint`, `available`, etc.) are `null` or `undefined`.  

### **5. Improve Error Handling Tests**  
- The tests handle database errors (`mockRejectedValue`), but additional cases should be tested:  
  - **Network failure**: Simulate network failures when calling the database.  
  - **Unhandled exceptions**: Use `try-catch` to verify error propagation.  

### **6. Verify Database Update Count**  
- `updateAssistantLimit` and `updateConfiguration` return status `200` even if `update()` modifies **zero** rows.  
- Test should ensure `update()` actually changes data before responding with success.  

### **7. Separate Concerns in Tests**  
- Some tests are checking multiple things (e.g., in `updateAssistantLimit`, it checks assistant count, database update, and error handling all in one test).  
- Each test should focus on **one** specific case to make failures easier to debug.

File: src/controllers/configuration.controller.js (hunk 1 above)
Comment (tirirote) [reply src/controllers/configuration.controller.js:50]: ok

File: src/models/models.js [hunk 3]
```diff
@@ -37,4 +37,4 @@ try {
 applyExtraSetup(sequelize);
 
 export { sequelize, Sequelize, models };
-export default models;
+export default models;
```
Comment (alvarobernal2412) [inline src/models/models.js]: This change should be discarded

File: src/utils/dates.js [hunk 4]
```diff
@@ -47,4 +47,4 @@ export function getDates(from, to, period, Wto, rules) {
   }
 
   throw new Error(`Invalid period type: ${period}`);
-}
+}
```
Comment (alvarobernal2412) [inline src/utils/dates.js]: This change should be discarded

**Review by alvarobernal2412 — CHANGES_REQUESTED** [review]
### **1. Validate Request Data Types**  
**Issue:** `updateAssistantLimit` should validate `req.params.limit` as a number.  

**Current Coverage:**  
- The test ensures `0` and `1` are handled but does not test invalid input like `"abc"`, `null`, or `undefined`.  

**Refactored Test (Using Parameterization):**  
```js
test.each([
  ['abc', 'Invalid limit value'],
  ['-5', 'Invalid limit value'],
  ['2.5', 'Invalid limit value'],
  [`${Number.MAX_SAFE_INTEGER + 1}`, 'Invalid limit value'],
  [null, 'Invalid limit value'],
  [undefined, 'Invalid limit value']
])('should return 400 for invalid limit value: %s', async (invalidLimit, expectedMessage) => {
  mockReq.params.limit = invalidLimit;
  
  await configurationController.updateAssistantLimit(mockReq, mockRes);
  
  expect(mockRes.status).toHaveBeenCalledWith(400);
  expect(mockRes.json).toHaveBeenCalledWith({ message: expectedMessage });
});
```  
**Action Taken:** Added parameterized test to cover all invalid values efficiently.  

---

### **2. Add Tests for Edge Cases**  
**Issue:**  
- `getConfiguration` should handle an empty array.  
- `updateAssistantLimit` should test extreme boundary values like `Number.MAX_SAFE_INTEGER`.  
- Case sensitivity of endpoints should be tested.  

**Refactored Tests:**  
```js
// Ensure getConfiguration handles an empty array
findAllSpy.mockResolvedValue([]);
await configurationController.getConfiguration(mockReq, mockRes);
expect(mockRes.status).toHaveBeenCalledWith(200);
expect(mockRes.json).toHaveBeenCalledWith([]);

// Ensure updateAssistantLimit handles large numbers
mockReq.params.limit = `${Number.MAX_SAFE_INTEGER}`;
await configurationController.updateAssistantLimit(mockReq, mockRes);
expect(mockRes.status).toHaveBeenCalledWith(200);
```  

**Action Taken:** Consolidated tests to cover edge cases without redundancy.  

---

### **3. Test for Unexpected Input Scenarios**  
**Issue:**  
- `updateConfiguration` should handle malformed JSON bodies.  

**Refactored Test:**  
```js
test.each([
  [{}, 'Invalid request body'],
  [null, 'Invalid request body'],
  [undefined, 'Invalid request body']
])('should return 400 for invalid request body: %s', async (invalidBody, expectedMessage) => {
  mockReq.body = invalidBody;
  
  await configurationController.updateConfiguration(mockReq, mockRes);
  
  expect(mockRes.status).toHaveBeenCalledWith(400);
  expect(mockRes.json).toHaveBeenCalledWith({ message: expectedMessage });
});
```  

**Action Taken:** Used parameterized tests for efficiency.  

---

### **4. Verify Database Update Count**  
**Issue:**  
- `updateConfiguration` should return an error if no records are updated.  

**Refactored Test:**  
```js
updateSpy.mockResolvedValue([0]);
await configurationController.updateConfiguration(mockReq, mockRes);
expect(mockRes.status).toHaveBeenCalledWith(400);
expect(mockRes.json).toHaveBeenCalledWith({ message: 'No records updated' });
```  

**Action Taken:** Ensured that `update()` modifying zero rows is correctly handled.  

---
## **5. Ensure `getAssistantLimit` Handles Null or Undefined Limits**  
### **Issue:**  
The test assumes `getAssistantLimit` always returns a valid `limit`, but if the database contains a `null` or `undefined` value, the response should be handled correctly.  

### **Current Coverage:**  
- Missing a test where `findOne()` resolves with `{ limit: null }`.  

### **Required Changes:**  
✅ Add test for **null assistant limit**:  
```js
findOneSpy.mockResolvedValue({ dataValues: { limit: null } });

await configurationController.getAssistantLimit(mockReq, mockRes);

expect(mockRes.status).toHaveBeenCalledWith(400);
expect(mockRes.json).toHaveBeenCalledWith({ message: 'Assistant limit is not set' });
```

✅ Add test for **undefined assistant limit**:  
```js
findOneSpy.mockResolvedValue({ dataValues: {} });

await configurationController.getAssistantLimit(mockReq, mockRes);

expect(mockRes.status).toHaveBeenCalledWith(400);
expect(mockRes.json).toHaveBeenCalledWith({ message: 'Assistant limit is not set' });
```

File: src/utils/dates.js (hunk 4 above)
Comment (tirirote) [reply src/utils/dates.js]: done

File: src/models/models.js (hunk 3 above)
Comment (tirirote) [reply src/models/models.js]: outdated, but done

File: src/controllers/configuration.controller.js
Comment (tirirote) [reply src/controllers/configuration.controller.js:1]: I've actually done the test from scratch again

**Comment by tirirote** [comment]
> Here are some areas for improvement in the testing of the controller:
> 
> ### **1. Improve Mocking of `updateConfigurationsCache`**
> * The test for `updateConfiguration` does not correctly invoke `updateConfigurationsCache`.
> * The function is referenced but never executed (`await updateConfigurationsCache;` should be `await updateConfigurationsCache();`).
> * The test should ensure this function is actually called.
> 
> ### **2. Validate Request Data Types**
> * In `updateAssistantLimit`, `req.params.limit` is a string, but it should be validated as a number.
> * Convert and validate `limit` before performing operations:
>   ```js
>   const limit = Number(req.params.limit);
>   if (isNaN(limit)) { /* return 400 error */ }
>   ```
> * The tests should check for invalid input cases (e.g., `"abc"` as a limit).
> 
> ### **3. Add Tests for Edge Cases**
> * **Empty Database Cases**:
>   
>   * What if `findAll()` in `getConfiguration` returns an empty array?
>   * The test cases assume configurations always exist.
> * **Boundary Values for `limit`**:
>   
>   * Test cases should check limits of `1`, `2`, `Number.MAX_SAFE_INTEGER`, etc.
> * **Case Sensitivity**:
>   
>   * What if an endpoint is requested in a different case (`/API/TEST` vs `/api/test`)?
> 
> ### **4. Test for Unexpected Input Scenarios**
> * **Invalid JSON Body**:
>   
>   * What happens if the request body is malformed or missing required fields?
>   * Example test:
>     ```js
>     mockReq = { body: {} }; 
>     await configurationController.updateConfiguration(mockReq, mockRes);
>     expect(mockRes.status).toHaveBeenCalledWith(400);
>     ```
> * **Null or Undefined Parameters**:
>   
>   * Ensure tests cover cases where parameters (`endpoint`, `available`, etc.) are `null` or `undefined`.
> 
> ### **5. Improve Error Handling Tests**
> * The tests handle database errors (`mockRejectedValue`), but additional cases should be tested:
>   
>   * **Network failure**: Simulate network failures when calling the database.
>   * **Unhandled exceptions**: Use `try-catch` to verify error propagation.
> 
> ### **6. Verify Database Update Count**
> * `updateAssistantLimit` and `updateConfiguration` return status `200` even if `update()` modifies **zero** rows.
> * Test should ensure `update()` actually changes data before responding with success.
> 
> ### **7. Separate Concerns in Tests**
> * Some tests are checking multiple things (e.g., in `updateAssistantLimit`, it checks assistant count, database update, and error handling all in one test).
> * Each test should focus on **one** specific case to make failures easier to debug.

**Review by alvarobernal2412 — APPROVED** [review]
